### PR TITLE
Implemented import command and allow duplicate spending

### PR DESCRIPTION
**Pull request by jonathantjendana**

### CODE REVIEW

**Comment by jonathantjendana** [comment]
#12

File: src/main/java/seedu/moneygowhere/model/spending/SpendingList.java [hunk 1]
```diff
@@ -8,22 +8,13 @@
...
-import seedu.moneygowhere.model.spending.exceptions.DuplicateSpendingException;
 import seedu.moneygowhere.model.spending.exceptions.SpendingNotFoundException;
 
 /**
- * A list of spendings that enforces uniqueness between its elements and does not allow nulls.
- * A Spending is considered unique by comparing using {@code Spending#isSameSpending(Spending)}.
- * As such, adding and updating of spending uses Spending#isSameSpending(Spending) for equality so as to ensure that the
- * Spending being added or updated is unique in terms of identity in the UniqueSpendingList.
- * However, the removal of a Spending uses Spending#equals(Object) so
- * as to ensure that the Spending with exactly the same fields will be removed.
```
Comment (Nanosync) [inline src/main/java/seedu/moneygowhere/model/spending/SpendingList.java:20]: Could removing this have any side effects when removing an existing spending?

File: src/main/java/seedu/moneygowhere/logic/commands/ImportCommand.java [hunk 2]
```diff
@@ -11,11 +35,116 @@
...
+            return new CommandResult(String.format(MESSAGE_SUCCESS, count));
+        }
+
+    }
+
+    /**
+     * Reads a csv file from the specified file path.
+     * @param path
+     * @return
+     * @throws ParseException
```
Comment (Nanosync) [inline src/main/java/seedu/moneygowhere/logic/commands/ImportCommand.java]: Need to specify these fields for javadoc

File: src/main/java/seedu/moneygowhere/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -129,8 +132,8 @@ public static Cost parseCost(String cost) throws ParseException {
      * Parses a {@code String remark} into an {@code remark}.
      * Leading and trailing whitespaces will be trimmed.
      */
-    public static Remark parseRemark(String remark) {
-        requireNonNull(remark);
+    public static Remark parseRemark(String remark) throws ParseException {
+        // requireNonNull(remark);
```
Comment (Nanosync) [inline src/main/java/seedu/moneygowhere/logic/parser/ParserUtil.java:137]: why was this change reversed?

File: src/test/java/seedu/moneygowhere/logic/parser/SpendingBookParserTest.java [hunk 4]
```diff
@@ -119,8 +119,8 @@ public void parseCommand_showBudget() throws Exception {
 
     @Test
     public void parseCommand_import() throws Exception {
-        assertTrue(parser.parseCommand(ImportCommand.COMMAND_WORD) instanceof ImportCommand);
-        assertTrue(parser.parseCommand(ImportCommand.COMMAND_WORD + " 3") instanceof ImportCommand);
+        assertTrue(parser.parseCommand(ImportCommand.COMMAND_WORD + " p/src/test/data/invalidDateSpending.csv")
```
Comment (Nanosync) [inline src/test/java/seedu/moneygowhere/logic/parser/SpendingBookParserTest.java]: use `PREFIX_PATH` for "p/"

File: src/main/java/seedu/moneygowhere/model/path/FilePath.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import seedu.moneygowhere.commons.util.FileUtil;
+
+/**
+ * Represents a file path.
+ */
+public class FilePath {
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "File path should not contain any of the following characters: < > ? [ ] : | * .  \n"
+                    + "It should not not contain more than 218 characters and it should be an existing csv file.";
```
Comment (Nanosync) [inline src/main/java/seedu/moneygowhere/model/path/FilePath.java]: **The file** should not contain more than 218 characters

File: src/main/java/seedu/moneygowhere/commons/util/JsonUtil.java [hunk 6]
```diff
@@ -139,5 +139,4 @@ private Level getLoggingLevel(String loggingLevelString) {
             return Level.class;
         }
     }
-
```
Comment (Nanosync) [inline src/main/java/seedu/moneygowhere/commons/util/JsonUtil.java]: Accidental change

File: src/main/java/seedu/moneygowhere/model/path/FilePath.java (hunk 5 above)
Comment (jonathantjendana) [reply src/main/java/seedu/moneygowhere/model/path/FilePath.java]: Why? The file can contain more than 218 characters, the path can’t

Comment (Nanosync) [reply src/main/java/seedu/moneygowhere/model/path/FilePath.java]: Grammar issue, can probably replace `it` to  `and not contain more than 218 characters`

File: src/main/java/seedu/moneygowhere/logic/parser/ParserUtil.java (hunk 3 above)
Comment (jonathantjendana) [reply src/main/java/seedu/moneygowhere/logic/parser/ParserUtil.java:137]: to allow duplicate

File: src/main/java/seedu/moneygowhere/model/spending/SpendingList.java (hunk 1 above)
Comment (jonathantjendana) [reply src/main/java/seedu/moneygowhere/model/spending/SpendingList.java:20]: nope

File: src/main/java/seedu/moneygowhere/logic/commands/ImportCommand.java [hunk 7]
```diff
@@ -11,11 +35,116 @@
...
+        } else {
+            return new CommandResult(String.format(MESSAGE_SUCCESS, count));
+        }
+
+    }
+
+    /**
+     * Reads a csv file from the specified file path.
+     * @param path
+     * @return Pair
```
Comment (austinsantoso) [inline src/main/java/seedu/moneygowhere/logic/commands/ImportCommand.java]: needs more descriptions

File: src/main/java/seedu/moneygowhere/logic/parser/ParserUtil.java (hunk 3 above)
Comment (austinsantoso) [reply src/main/java/seedu/moneygowhere/logic/parser/ParserUtil.java:137]: approve

File: src/main/java/seedu/moneygowhere/logic/commands/ImportCommand.java [hunk 8]
```diff
@@ -11,11 +35,116 @@
...
+    private final FilePath fullFilePath;
+
+    public ImportCommand(FilePath filePath) {
+        requireNonNull(filePath);
+        fullFilePath = filePath;
+    }
+
+    /**
+     * Converts a map of objects into a Spending object.
+     * @param map
```
Comment (austinsantoso) [inline src/main/java/seedu/moneygowhere/logic/commands/ImportCommand.java]: i think this needs more descriptions

File: src/main/java/seedu/moneygowhere/logic/commands/ImportCommand.java [hunk 9]
```diff
@@ -11,11 +35,116 @@
...
+            return new CommandResult(String.format(MESSAGE_SUCCESS, count));
+        }
+
+    }
+
+    /**
+     * Reads a csv file from the specified file path.
+     * @param path
+     * @return Pair
+     * @throws ParseException
```
Comment (austinsantoso) [inline src/main/java/seedu/moneygowhere/logic/commands/ImportCommand.java]: i think this needs more descriptions

File: src/main/java/seedu/moneygowhere/logic/commands/ImportCommand.java [hunk 10]
```diff
@@ -11,11 +35,116 @@
...
+
+    public ImportCommand(FilePath filePath) {
+        requireNonNull(filePath);
+        fullFilePath = filePath;
+    }
+
+    /**
+     * Converts a map of objects into a Spending object.
+     * @param map
+     * @return Spending
```
Comment (austinsantoso) [inline src/main/java/seedu/moneygowhere/logic/commands/ImportCommand.java]: i think this needs more descriptions

**Review by austinsantoso — CHANGES_REQUESTED** [review]

**Review by austinsantoso — APPROVED** [review]
lgtm

**Pull request merged**